src/languageclient.rs

```diff
@@ -720,15 +720,18 @@ impl State {
             return Ok(result);
         }
 
+        let hover_preview: u8 = self.eval("get(g:, 'LanguageClient_hoverPreview', 1)")?;
+
         let hover: Option<Hover> = serde_json::from_value(result.clone())?;
         if let Some(hover) = hover {
-            if hover.lines_len() <= 1 {
-                self.echo(hover.to_string())?;
-            } else {
+            if hover_preview == 2 || (hover_preview == 1 && hover.lines_len() > 1) {
                 self.preview(&hover.to_display())?;
+            } else {
+                self.echo(hover.to_string().lines().last().unwrap_or_default())?;
```

Comment:
Instead of showing only last line, use `echo_ellipsis` to show as much information as possible.

Reply:
I agree that `echo_ellipsis` is a better solution so I switched to that. It's still not perfect; however, since the most relevant part of the hover output is usually the type signature and all of the language servers that I have tried (`rls` and `cquery` included) put the type signature at the end of the hover output.

Ideally I would be able to use the `signatureHelp` request to get this info, but not all servers support it 😞